utils/iam_dataset.py

```diff
@@ -0,0 +1,298 @@
+import cv2
+import json
+import numpy as np
+import pandas as pd
+import zipfile
+import matplotlib.pyplot as plt
+
+from mxnet.gluon.data import dataset
+from mxnet import nd
+
+class IAMDataset(dataset.ArrayDataset):
+    """ The IAMDataset provides images of handwritten passages written by multiple
+    individuals. The data is available at http://www.fki.inf.unibe.ch
+
+    The passages can be parsed into separate words, lines, or the whole form.
+    The dataset should be separated into writer independent training and testing sets.
+
+    Parameters
+    ----------
+    parse_method: str, Required
+        To select the method of parsing the images of the passage
+        Available options: [form, line, word]
+
+    credentials: (str, str), Default None 
+        Your (username, password) for the IAM dataset. Register at
+        http://www.fki.inf.unibe.ch/DBs/iamDB/iLogin/index.php
+        By default, IAMDataset will read it from credentials.json
+    
+    root: str, default: dataset/iamdataset
+        Location to save the database
+
+    train: bool, default True
+        Whether to load the training or testing set of writers.
+
+    output_data_type: str, default text
+        What type of data you want as an output: Text or bounding box.
+        Available options are: [text, bb]
+     
+    output_parse_method: str, default None
+        If the bounding box (bb) was selected as an output_data_type, 
+        this parameter can select which bb you want to obtain.
+        Available options: [form, line, word]
+
+    transform: function, default None
+        A user defined callback that transforms each sample. For example:
+    ::
+        transform=lambda data, label: (data.astype(np.float32)/255, label)
+
+    """
+    MAX_IMAGE_SIZE_FORM = (1120, 800)
+    def __init__(self, parse_method, credentials=None,
+                 root=os.path.join('dataset', 'iamdataset'), 
+                 train=True, output_data="text",
+                 output_parse_method=None, transform=None):
+
+        _parse_methods = ["form", "line", "word"]
+        error_message = "{} is not a possible parsing method: {}".format(
+            parse_method, _parse_methods)
+        assert parse_method in _parse_methods, error_message
+        self._parse_method = parse_method
+        url_partial = "http://www.fki.inf.unibe.ch/DBs/iamDB/data/{data_type}/{filename}.tgz"
+        if self._parse_method == "form":
+            self._data_urls = [url_partial.format(data_type="forms", filename="forms" + a) for a in ["A-D", "E-H", "I-Z"]]
+        elif self._parse_method == "line":
+            self._data_urls = [url_partial.format(data_type="lines", filename="lines")]
+        elif self._parse_method == "word":
+            self._data_urls = [url_partial.format(data_type="words", filename="words")]
+        self._xml_url = "http://www.fki.inf.unibe.ch/DBs/iamDB/data/xml/xml.tgz"
+
+        if credentials == None:
+            if os.path.isfile("credentials.json"):
```

Comment:
`os.path.join(os.path.dirname(__file__), '..', 'credentials.json')` would be safer, so that you can execute from anywhere, just need to make sure the `credentials.json` file is at the root of your repo.